README.md
AncientAtoms
============

This is the ancient, entirely unsupported, written-in-Fortran version
of Atoms

The Makefiles work on a computer running Ubuntu 13.04 using gfortran
from GCC 4.7.3.  The resulting executable has been shown to run to
completion for one test case.  It has not been tested with any other
crystal data, on any other computer or with any other compiler.

This source code is being placed on GitHub as a courtesy.  I will
provide no support of any kind for the use of this software.  I will
provide no advice for how to compile this software.  I will fix no
bugs.  I will answer no questions.  Indeed, any questions or comments
you have on this version of the program are likely to be silently
ignored.  This is obsolete software.

I hope the preceding paragraph is clear.

The entire functionality of this program was reimplemented many years
ago and has since been integrated into my Artemis program, a part of
[Demeter](http://bruceravel.github.io/demeter/).


headers/history
c    1.0    got it working
c    2.0    satisfied copyright criteria in cluster expansion routines
c    2.1    added indexing (8/92)
c    2.2    added absorption length (9/92)
c    2.3    improved handling of spc grp notation, fixed bcc bug,
c           fixed floating point logic (1/93)
c    2.4    handle bases, atom list option, shift (3/93), mcmaster
c           correction (4/93)
c  approaching 3 too quickly, switch to hundredths
c    2.41   mcmaster correction corrected, i0 and self absorption
c           corrections, new cluster expansion routine, overfull
c           unit cell, simple dopant, more error checking on input, 
c           multiple title lines, untab, shift warning and
c           version number in feff.inp (6/93)
c    2.42   modularized in anticipation of dafs applications (6/93)
c           general housekeeping, tags, improved dopants, core keyword 
c           required, case insensitivity in code, geom.dat + subshell 
c           sorting, self-absorption bug fixed (8/93 and 10/93) 
c    2.42.a lower case and true robustness, revamped input parsing for 
c           better error checking, handle overlapping atoms better, hex 
c           C, V symmetry groups, corrected tri/hex bug in atinpt, no 
c           mcmaster for Z<13, diagnostic messages from input file (1/94)
c           corrected bug in dopants,
c           more housekeeping (2/94) calculate delta mu from core rather
c           than total mu and reinstate mcmaster for Z<13 (3/94)
c           silly bug in atinpt that trashed nonorthoganal groups fixed,
c           also improved error messages (3/94)
c    2.42.b fixed mishandling of rhombohedral groups (6/94)
c    2.42.c pass parameters as arguments for easier hacking, had to pass
c           some memory space to certain subroutines, fixed a bug 
c           in subroutine ref that was affecting geout, 72 character
c           names for output files, pass io number for feff.inp, 
c           l1&l2, move origin into readin (8/94)
c    2.42.d contains full fuctionality of 2.43.c except for dafs apps (1/95)
c    2.42.e contains full fuctionality of 2.43.d except for dafs apps (2/95)
c    2.42.f contains full fuctionality of 2.43.f except for dafs apps 
c     & up   parallel development to 2.43 without dafs apps
c
c    2.43   tabulate chantler's data for interpolation, wrote fcal -- 
c           an engine for using chantler's data, wrote a0.f for reading
c           chantler and cromer-mann and making a0, write a0 as a five 
c           column file suitable for feffit, no energy corrections (8/94)
c    2.43.b new a0 filename, allow neg. q's, fix null atoms handling, 
c           fixed mishandling of trigonal (not rhomb.) groups (9/94)  
c           added warning message for low symmetry cells, added my name 
c           to feff.inp and run time messages(11/94), fix some error
c           messages (1/95)
c    2.43.c full sasaki tables, fixed espilon bug in multip, added p1 
c           option  (1/95)
c    2.43.d after UWXAFS 3.0 release -- fixed bug with dmax < min distance
c           (2/95)
c    skip subversion e
c    2.43.f fixed bug in self-absorption calculation, added krypton to I0,
c           geom.dat and feff.dat write same number of atoms (4/95)
c    2.43.g no file name conflicts, much improved internal documentation
c           (7/95)
c
c    2.44. & 2.45.  handles different settings of low symmetry crystals,
c                   improve handling of schoenflies notation, better error
c                   messages for mcmaster calculations (8/95), compile time
c                   switch for stdin/stdout (9/95), typo in syschk, in
c                   multip minor bug re. imult for atom with coord=1 
c                   edge energy in feff.inp, allow L1&l2 (10/95)
c    2.44 & up parallels 2.45 without dafs apps
c    2.45.a fixed bug in f.p. comparisons in indexing, five digits in 
c           feff.inp, geom.dat no longer depends on scratch file (11/95)
c    2.45.b central atom gets tag, reflect.dat (12/95)
c
c    2.46 & up parallels 2.47 without dafs apps and extended atoms list
c    2.47   tighten modularity & combine variables into arrays, fix problem
c           with nofx and polyft by explicitly carrying around misc.f file,
c           information for fluorescence normalization,  move atoms list 
c           parsing to external subroutine to allow expanded atoms list, 
c           rmax card to 5 digits + 0.00001, fixed potential numerical 
c           problem in metric (2/96) made groups module (3/96) angles in  
c           monoclinic settings, 
c    2.4?.b rework makefile using uninclude perl script and unzipped
c           modules, added xanes card 
c    2.4?.c added keywords "feff8" and "correction", added optional
c           feff8 output (BR Jan 16 1998)
c
c    to do: new crystal, fix monoclinic angles
c
c    2.50   change structure of source tree, use libraries and include 
c           files in a more sensible manner, (BR Mar  2 1998) (goals
c           for 2.50pre1: modularize crystl+groups+clustr, remove logic()
c
c    changes in functionality from versions before 2.50
c    1. no longer support more than one job per file
c    2. consistent message syntax for post-processing purposes
c
c    goals for 2.50pre2: clean up remaining code, allow spcgrp=ispa,
c                        end keyord wade through code cleaning up
c                        run-time messages
c    goals for 2.50pre3: begin work on monoclinic
c    other goals for 2.50: revise document, write crystl user guide,
c       more output types, see wishlist


atom.f
      program atoms
      implicit real(a-h,o-z)
      implicit integer(i-n)
      include 'version.h'
c----------------------------------------------------------------------
c             copyright 1998  Bruce Ravel
c        copyright 1993-1997  University of Washington
c             written by      Bruce Ravel
c                             Ceramics Division
c                             National Institute of Standards and Technology
c                             Bldg 223, Room 329
c                             Gaithersburg, MD 20899
c
c                  phone      (301) 975-5759
c                  fax        (301) 975-5334
c                  e-mail     bruce.ravel@nist.gov
c                          or ravel@phys.washington.edu
c           please use email for communication with the author
c
c  insert permissions notice here
c----------------------------------------------------------------------
c      brief description of the code:
c
c  atoms writes a list of atomic coordinates for any crystal given its
c  crystallographic information. the list will be sorted by radial
c  distance from an atom chosen as the central atom. atoms also estimates
c  the bulk absorption and density of the material and various corrections
c  to xafs data due to experimental effects.
c----------------------------------------------------------------------
c  comments blocks that follow:
c       glossary of variables
c       descriptions of runtime error codes
c       sample input file
c       version history
c
c  the code begins after the first occurance of this string:  %%%%
c  the first executable statement begins after its second occurance
c----------------------------------------------------------------------
c >>> glossary of variables
      include 'glossary'
c=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
c >>> runtime error codes
      include 'runtime'
c=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
c >>> sample input file:
      include 'sample'
c=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
c >>> history
      include 'history'
c----------------------------------------------------------------------
c%%%%
      include 'atparm.h'

c  stdout=.true. for reading from standard input and writing to standard
c  output
      logical stdout
      parameter (stdout=.false.)

c  expnd=.true. for reading expanded atoms list, false for normal atoms list
      logical expnd
      parameter (expnd=.false.)

c  vaxflg=.true. for compilation on a VMS machine, used for opening
c  files in such a way that VMS version numbers are used
      logical vaxflg
      parameter (vaxflg=.false.)

c  crystl.h contains all information relevant to a unit cell
      include 'crystl.h'
c  exafs.h contains information relevant to exafs calculations
      include 'exafs.h'
c  unit.h contains information over-full unit cell information
      include 'unit.h'

c  see glossary for descriptions of variables, see arrays.map for
c  description of logic, gasses, exafs
      parameter(m=8)
      character*2  elemnt(iat), test, vrsn*9
      character*10 inpgrp, noantg(iat)
      character*72 title(ntitx), outfil, afname, refile
      character*78 messg
      character*78 module, string
      logical      logic(nlogx)
      complex      anot(neptx)
      dimension    ngeom(ngeomx)
      dimension    atlis(natx,8), qvect(3)
      dimension    nrefl(3)


c  reserve some array space for use in modules cluster and output
      dimension    cltmp1(natx),  cltmp2(natx),  cltmp3(natx,m)
      logical      cltmp4(natx)
      character*10 optmp1(maxln)
      dimension    optmp2(maxln), optmp3(maxln), optmp4(maxln),
     $             optmp5(maxln), ioptp6(maxln), ioptp7(maxln)

c------------------------------------------------------------------------
c-- dafs stuff
c       dimension    f0(iat,ndopx), usqr(iat)
c       complex      fcore(neptx)
c c  from block data sasaki
c       parameter(nelem=92, ndatx=233)
c       common /fdat/ nfdata(nelem), engrd(nelem,ndatx),
c      $            fp(nelem,ndatx), fpp(nelem,ndatx)
c c  this is the crommer-mann common block
c       common /sk/ aa(1926)
c------------------------------------------------------------------------

c------------------------------------------------------------------------
c  formats needed at top level: version # (4000), line of '=' (4010),
c                 ... aN,(55-N)x ...
 4000 format('ATOMS ',a9,46x,'by Bruce Ravel')
 4010 format(75('='))

c%%%%
c---------------------------------------------------------------------
c  run time messages: lines of '=' and version number
c  set unit number of feff.inp
      if (.not.stdout) then
          ifeff = 2
          write(messg,4010)
          call messag(messg)
          write(messg,4000)vrsion
          call messag(messg)
          write(messg,4010)
          call messag(messg)
      else
c  this clause may trigger a compiler warning regarding there being no
c  possible path to this spot in the code.  this is an unavoidable
c  result of using F77 and no preprocessor and should pose no problem
c  during execution of the code.
          ifeff = 6
      endif

c---------------------------------------------------------------------
c  initialize some top-level variables
      vrsn  = vrsion
      test  = 'ab'
      logic(1) = .false.
      module = 'Atoms'

c 10    continue

c=====================================================================
c  atoms module 1:  initialize, read input file, error checking
c=====================================================================
c      print*,'beginning module 1'
      call readin(iat, natx, ntitx, ndopx, ngeomx, neptx, nlogx,
     $            ifeff, ntit, iatom, ibasis, iabs, isystm,
     $            iperm, ipt, iptful, idop, ngeom, iedge,
     $            nepts, nsites, nrefl, nnoan,
     $            vrsn, spcgrp, inpgrp, title, outfil, afname,
     $            refile, tag, noantg, edge, core, elemnt, dopant,
     $            x, y, z, cell, rmax, st, fullcl, atlis,
     $            percnt, gasses, qvect, egr, anot,
     $            logic, stdout, vaxflg, expnd)

      if (logic(1)) goto 99

c=====================================================================
c  atoms module 2:  decode space group, determine unit cell contents
c=====================================================================
      string = 'the crystl module'
      if (logic(20)) call positn(module, string)
      iunidb = 0
      if (logic(21)) icrydb = icrydb + 2**0
      call crystl(icrydb, iercry)

      if ( (.not.stdout).and.syserr ) then
          call messag(' ')
          call messag(' *** Warning ')
          do 20 i=1,nsysm
            call messag(sysmes(i))
 20       continue
          call messag('The calculation will be finished, but you '//
     $                'might want to edit your')
          call messag('crystallographic input data and try again.-')
          call messag(' ')
      endif

c=====================================================================
c  atoms module 3:  perform various calculations using mcmaster tables
c=====================================================================
      if (logic(28)) then
          string = 'the mcm module'
          if (logic(7)) then
              if (logic(20)) call positn(module, string)
              iexadb = 0
              if (logic(22)) iexadb = iexadb + 2**0
              if (logic(14)) iexadb = iexadb + 2**1
              if (logic(15)) iexadb = iexadb + 2**2
              if (logic(16)) iexadb = iexadb + 2**3
              call mcm(iexadb)
          endif
          if (.not.lfluo) logic(3)=.false.
      endif

c=====================================================================
c  atoms module 4:  construct unit.dat
c=====================================================================
      string = 'the unit module'
      if (logic(20)) call positn(module, string)
      iunidb = 0
      if (logic(23)) iunidb = iunidb + 2**0
      if (logic(8))  iunidb = iunidb + 2**1
      if (logic(9))  iunidb = iunidb + 2**2
      call unit(iunidb, ntit, title, vaxflg, ieruni)

c=====================================================================
c  atoms module 5:  construct structure factor for dafs applications
c=====================================================================
c       if (logic(10).or.logic(11)) then
c           string = 'the ascat module'
c           if (logic(20)) call positn(module, string)
c           call ascat(iat, ntitx, ndopx, neptx, nlogx,
c      $             nsites, ipt, idop, ntit, nepts, nrefl, nnoan,
c      $             edge, core, dopant, title, afname, refile, vrsn,
c      $             tag, noantg,
c      $             st, usqr, cell, percnt, qvect, egr, fcore, f0,
c      $             anot, logic, vaxflg)
c       endif

c=====================================================================
c  atoms module 6:  expand cluster around central atom
c=====================================================================
      if (logic(7)) then
          string = 'the clustr module'
          if (logic(20)) call positn(module, string)
          call clustr(iat,natx,ngeomx,nlogx,
     $            iabs,nsites,iperm,ipt,itot,ngeom,
     $            cell,trmtx,rmax,st,atlis,
     $            cltmp1,cltmp2,cltmp3,cltmp4,
     $            logic)

c=====================================================================
c  atoms module 7:  write feff.inp, geom.dat
c=====================================================================
          string = 'the output module'
          if (logic(20)) call positn(module, string)
          call output(iat, natx, ntitx, ndopx, ngeomx, maxln, nlogx,
     $            nexafs,
     $            ifeff, iabs, itot, ntit, idop, ngeom, imult,
     $            title, tag, edge, core, dopant, outfil, elemnt,
     $            vrsn, percnt, exafs, atlis,
     $            logic, vaxflg,
     $            optmp1,optmp2,optmp3,optmp4,optmp5,ioptp6,ioptp7)

c#####################################################################

c---------------------------------------------------------------------
c  run time message: output file and line of '='
          if ((logic(6)).and.(outfil.ne.'list')) then
              if (logic(13)) then
                  call messag('geom.dat overwritten.')
              else
                  call messag('geom.dat written.')
              endif
          endif
          if (.not.stdout) then
              ii = istrln(outfil)
              if (logic(12)) then
                  call messag(outfil(:ii)//' overwritten.')
              else
                  call messag('Output written to '//outfil(:ii))
              endif
          endif
      else
          call messag('not writing feff.inp.')
      endif
      if (.not.stdout) then
          write(messg,4010)
          call messag(messg)
      endif
c---------------------------------------------------------------------

c      if (.not.logic(1)) goto 10
99    continue
      stop
c  end main program atom
      end
